Keyframe Selection: Rework and add new selection methods

cbentejac opened the pull request:
## Description

This PR is a replacement for #1213 and aims at completely reworking the Keyframe Selection utility.

It removes the existing selection methods and replaces them with two selection methods:
- A "regular" selection method, which samples frames regularly over time with respect to some user-provided constraints (namely the maximum number of wanted keyframes, the minimum number of frames between two selected keyframes, and the maximum number of frames between two keyframes). 
This was already possible with the existing Keyframe Selection utility, but has been simplified to be more user-friendly.
- A "smart" selection method, which computes scores for each frame in the input sequence(s) and uses them to perform the final selection. Scores are based on the sharpness, motion and temporal position of the evaluated frame. 
The goal is to select a frame that is as sharp as possible with significant motion compared to the previously selected frame: consecutive frames should not be selected as keyframes if they do not contain enough motion, even if they are both very sharp. The level of motion that is deemed significant can be adjusted with parameters, as well as the minimum and maximum number of expected keyframes.

Additionally, debug options have been added:
- The scores computed for each frame can be exported as a CSV file;
- The motion vectors of each frame can be exported as .png images to be visualised, with the color indicating the motion direction and the intensity indicating its magnitude. 

The name of output frames can also be chosen between two options:
- The frames are named using their index number within the sequence like [00012, 00253, 00498] (default);
- The frames are named with consecutive frame numbers like [00000, 00001, 00002].

Videos, sequences of images and camera rigs are supported as inputs.

The corresponding Meshroom pull request can be found here: https://github.com/alicevision/Meshroom/pull/1880


## Features list

- [x] Remove the existing selection methods;
- [x] Add a simple selection method that samples frames regularly;
- [x] Add a smart selection method that evaluates the sharpness and optical flow of each frame to select only the most relevant;
- [x] Add debug options that allow to write scores to a CSV file and visualise the optical flow frames;
- [x] Add an option that allows choose how the ouptut frames will be named.

## Implementation remarks

- The sharpness of every frame is evaluated with a sliding window instead of as a whole. The goal is to ensure that an image will not be thrown away if it contains a sharp object while the rest of the image is blurry.
N.B: The most obvious performance improvement that could be done would concern the sliding window: it currently slides pixel by pixel across all the image, which is very time-consuming but doesn't necessarily bring in more information than if it was sliding faster.
- The optical flow is computed for cells within the image, giving for each of these cells an average pixel displacement within that cell. The median flow value of all the cells is then selected as the motion score.
- Once the sharpness and optical flow scores have been computed, the accumulated motion across frames is computed, and using a threshold, subsequences are identified. Within each subsequence, a keyframe is selected based on both its sharpness
and its position in the subsequence (frames located at the middle have an advantage over frames located on the subsequence's borders). The goal is to select frames that show a significant difference in their motion, are not too close to each other in time, and are as sharp as possible.
- By default, scores are computed on downscaled frames. The downscale levels for the sharpness and motion computations may differ depending on the set parameters.

### Known limitations

- The "Orientation" exif tag is not properly propagated for PNG outputs, as PNG metadata differs from JPG/EXR files.
- The "PixelAspectRatio" information is not propagated to JPG outputs: for unknown reasons, writing the PixelAspectRatio in JPG the same way it is for PNG/EXR outputs leads to some mistakes when it is bigger than 1 (writing a PixelAspectRatio of 2 leads to a JPG image with a PixelAspectRatio of 0.5).

mugulmd commented on src/aliceVision/keyframe/KeyframeSelector.cpp:
```diff
@@ -25,452 +25,711 @@ namespace keyframe {
+ *                  There will be no resizing if this parameter is set to 0
+ * @return An OpenCV Mat object containing the image
+ */
+cv::Mat readImage(dataio::FeedProvider &feed, std::size_t width = 0)
```
maybe place the declaration and documentation in header file ?

cbentejac replied:
I added it to the KeyframeSelector class as a private member. Done.

mugulmd commented on src/aliceVision/keyframe/KeyframeSelector.cpp:
```diff
@@ -25,452 +25,711 @@ namespace keyframe {
+    std::string currentImgName;
+
+    if (!feed.readImage(image, queryIntrinsics, currentImgName, hasIntrinsics)) {
+        ALICEVISION_LOG_ERROR("Cannot read frame '" << currentImgName << "'!");
```
maybe use `ALICEVISION_THROW_ERROR` to have only one line instead of two ? (not sure the behavior is exactly the same though)

mugulmd commented on src/aliceVision/keyframe/KeyframeSelector.cpp:
```diff
@@ -25,452 +25,711 @@ namespace keyframe {
+    _selectedKeyframes.clear();
+    _selectedFrames.clear();
+
+    /**
```
this kind of high-level explanation could also be in the function documentation, this way it would be accessible from the documentation generated by doxygen

cbentejac replied:
Done.

mugulmd commented on src/aliceVision/keyframe/KeyframeSelector.cpp:
```diff
@@ -25,452 +25,711 @@ namespace keyframe {
-            maxSharpness = _framesData[index].avgSharpness;
-          }
+    // Step 3: for each subsequence, find the keyframe
+    for (std::size_t i = 1; i < subsequenceLimits.size(); i++) {
```
`++i` instead of `i++` (same in several other places below)

cbentejac replied:
Done.

mugulmd commented on src/aliceVision/keyframe/KeyframeSelector.cpp:
```diff
@@ -25,452 +25,711 @@ namespace keyframe {
+            filenameSS << std::setw(5) << std::setfill('0') << pos << ".exr";
+            const auto filepath = (processedOutputFolder / fs::path(filenameSS.str())).string();
+
+            std::unique_ptr<oiio::ImageOutput> out(oiio::ImageOutput::create(filepath));
```
you can use `image::writeImage`

cbentejac replied:
Done.

mugulmd commented on src/aliceVision/keyframe/KeyframeSelector.hpp:
```diff
@@ -32,334 +32,206 @@ namespace oiio = OIIO;
+     * @param[in] mmFocals focal in millimeters for each camera
+     * @return true if all the selected keyframes were successfully written, false otherwise
+     */
+    bool writeSelection(const std::vector<std::string>& brands, const std::vector<std::string>& models,
```
this method could be `const`

cbentejac replied:
Done.

mugulmd commented:
Would be great to be able to export in other formats (jpg, png, etc.)

mugulmd commented:
There seem to be a problem when writing the selected frames: images are brighter than they should be (maybe a gamma is applied ?)

simogasp commented on src/aliceVision/keyframe/KeyframeSelector.cpp:
```diff
@@ -25,452 +25,711 @@ namespace keyframe {
+        const auto& path = _mediaPaths.at(mediaIndex);
+
+        // Create a feed provider per mediaPaths
+        feeds.emplace_back(new dataio::FeedProvider(path));
```
it should be safer to use `make_unique()` https://stackoverflow.com/a/15784813  @fabiencastan 
```suggestion
        feeds.push_back(std::make_unique<dataio::FeedProvider>(path));
```

simogasp commented on src/aliceVision/keyframe/KeyframeSelector.cpp:
```diff
@@ -25,452 +25,711 @@ namespace keyframe {
+
+    /* Starts at 1 because the first frame's motion score will be -1.
+     * Ends at sequenceSize - 1 to ensure the last frame cannot be pushed twice. */
+    for (std::size_t i = 1; i < sequenceSize -1; i++) {
```
```suggestion
    for (std::size_t i = 1; i < sequenceSize -1; ++i) {
```

cbentejac replied:
Done.

simogasp commented on src/aliceVision/keyframe/KeyframeSelector.cpp:
```diff
@@ -25,452 +25,711 @@ namespace keyframe {
+        const auto& path = _mediaPaths.at(mediaIndex);
+
+        // Create a feed provider per mediaPaths
+        feeds.emplace_back(new dataio::FeedProvider(path));
```
```suggestion
        feeds.push_back(std::make_unique<dataio::FeedProvider>(path));
```

cbentejac replied:
Done.

simogasp commented on src/aliceVision/keyframe/KeyframeSelector.cpp:
```diff
@@ -25,452 +25,711 @@ namespace keyframe {
+            }
+
+            // Compute sharpness
+            double sharpness = computeSharpness(currentMat, sharpnessWindowSize);
```
```suggestion
            const double sharpness = computeSharpness(currentMat, sharpnessWindowSize);
```

cbentejac replied:
Done.

simogasp commented on src/aliceVision/keyframe/KeyframeSelector.cpp:
```diff
@@ -25,452 +25,711 @@ namespace keyframe {
+            tr = squaredSum.at<double>(y, x + windowSize);
+            bl = squaredSum.at<double>(y + windowSize, x);
+            br = squaredSum.at<double>(y + windowSize, x + windowSize);
+            double s2 = br + tl - tr - bl;
+
+            double std2 = std::sqrt((s2 - (s1 * s1) / totalCount) / totalCount);
```
```suggestion
            const double s2 = br + tl - tr - bl;

            const double std2 = std::sqrt((s2 - (s1 * s1) / totalCount) / totalCount);
```

cbentejac replied:
Done.

simogasp commented on src/aliceVision/keyframe/KeyframeSelector.cpp:
```diff
@@ -25,452 +25,711 @@ namespace keyframe {
+            }
+        }
+
+        for (auto pos : _selectedKeyframes) {
```
```suggestion
        for (const auto pos : _selectedKeyframes) {
```

cbentejac replied:
Done.

simogasp commented:
It would be also nice to add a `README.md` in `keyframe` like for the other modules to have some documentation of the methods implemented, references if any etc...

cbentejac commented:
> Would be great to be able to export in other formats (jpg, png, etc.)

Added! Currently supported formats are exr, jpg and png.

> There seem to be a problem when writing the selected frames: images are brighter than they should be (maybe a gamma is applied ?)

There was indeed an issue with the colorspace for input videos, it is now fixed!

cbentejac commented:
> It would be also nice to add a `README.md` in `keyframe` like for the other modules to have some documentation of the methods implemented, references if any etc...

All done!

simogasp commented on src/aliceVision/keyframe/README.md:
```diff
@@ -0,0 +1,114 @@
+## Regular selection method
+
+The regular selection samples frames regularly over time with respect to some user-provided constraints:
+- `minFrameStep`: the minimum number of frames between two selected keyframes;
+- `maxFrameStep`: the maximum number of frames between two selected keyframes;
```
not clear to me what the role of the two parameters is: if it is a regular sampling of the frames (e.g. 1 frame every 5) why do we need 2 parameters? Or there is still a sharpness score influencing the choice of the frame? Like take  the sharpest frame that is at least `minFrameStep` far from the previous and at most `maxFrameStep` far from the previous one. Is it like this?

cbentejac replied:
There is no influence of the sharpness or the optical flow in the regular method, it really is just a regular sampling that can be modulated. 
`maxFrameStep` is mostly useful when you want to combine a relatively strict sampling with a maximum number of output frames. It is basically here to take precedence over `maxNbOutFrames`.

The expected behaviour for the regular selection depending on the set parameters can be summed up as follows:

- If only `minFrameStep` is set, the whole sequence will be sampled and a keyframe will be selected every `minFrameStep`. E.g: your sequence has 2000 frames, `minFrameStep = 100`, you will get 21 keyframes: with exactly 100 frames between them.

- If `minFrameStep` and `maxNbOutFrames` are set, you will never have less than `minFrameStep` between your keyframes, but you might have more for the whole sequence to be sampled while respecting `maxNbOutFrames`. E.g: your sequence has 2000 frames, `minFrameStep = 100` and `maxNbOutFrames = 10`: you will get 10 keyframes with 222 frames between them, so both `maxNbOutFrames` and `minFrameStep` are respected. If your sequence has 500 frames and `minFrameStep = 100` / `maxNbOutFrames = 10`, you will get 6 keyframes with 100 frames between them. No matter the value of the parameters, the entire sequence will be sampled.

- If `minFrameStep` and `maxFrameStep` are both set but `maxNbOutFrames` isn't, then the step between two keyframes will be exactly between `minFrameStep` and `maxFrameStep`. If `minFrameStep = 100` and `maxFrameStep = 200` without other constraints, it is equivalent to setting `minFrameStep = 150` (so, not the most useful use-case).

- If `minFrameStep`, `maxFrameStep` and `maxNbOutFrames` are all set, then `maxFrameStep` prevents the step between two frames from increasing too much to respect `maxNbOutFrames`. If we take again the example of the sequence with 2000 frames, if we have `minFrameStep = 100`, `maxFrameStep = 150` and `maxNbOutFrames = 10`, then we will get 10 keyframes with 150 frames between each, and the sampling will stop before reaching the end of the sequence so that all the constraints are respected. In the same example, if `maxFrameStep = 300`, then we will end up with 10 keyframes with 222 frames between them, and the whole sequence will be sampled.
`maxFrameStep` **ensures** that there will not be way more frames between two keyframes that we might want. 

I hope this is clear enough (and if you have any suggestion to make that clearer in the README, please let me know)!

simogasp replied:
Thanks for the explanation! :+1: 
Ok, it's not that trivial either then :-)

I think most of the users would be interested in the two basic settings:
1. taking one frame every `minFrameStep` all along the video
2. taking `maxNbOutFrames` equally spaced along the video.

Then all the others are more fine-tuning settings for advanced needs.
So at least in the documentation, we can highlight those two primary cases and then make a sub-section (`Advanced sampling`) where we explain all the other combinations and possibilities to tune the sampling further.
And since you already took the time to kindly write it you can re-use what you write with the examples because it is much more clear.

What do you think?

cbentejac replied:
Sounds good to me! I updated the `README.md` accordingly with these clarifications and examples.
Thanks for the feedback!

mugulmd commented:
An option that could be interesting for users would be to automatically rename the exported frames with their frame number within the output sequence.

Example:
- I want to do a keyframe selection on a sequence with 1000 frames
- the automatic process selects frames 15, 294, 600 and 825
- by default they will be saved as 0015.exr, 0294.exr, 0600.exr, 0825.exr
- with this option they would be respectively named 0000.exr, 0001.exr, 0002.exr, 0003.exr.

I don't know if it makes sense to add that feature here or if we should juste let users take care of the renaming with some other software, but in my case I know I would definitely use it 😅

cbentejac commented:
> An option that could be interesting for users would be to automatically rename the exported frames with their frame number within the output sequence.
> 
> Example:
> 
> * I want to do a keyframe selection on a sequence with 1000 frames
> * the automatic process selects frames 15, 294, 600 and 825
> * by default they will be saved as 0015.exr, 0294.exr, 0600.exr, 0825.exr
> * with this option they would be respectively named 0000.exr, 0001.exr, 0002.exr, 0003.exr.
> 
> I don't know if it makes sense to add that feature here or if we should juste let users take care of the renaming with some other software, but in my case I know I would definitely use it 😅

This was very easy to add and not costly at all, so I added an option to do what you described (`renameKeyframes`, which is disabled by default). Thanks for the suggestion!

mugulmd commented:
Might be a good idea to add some info logs (using `ALICEVISION_LOG_INFO`) to keep track of the process: 
I just tested it on a 2600 frames sequence and it took about 40 minutes, which without any logging can seem quite suspicious to users (as they might want at least a rough idea of how fast the process is going and how long it is going to last).

cbentejac commented:
> Might be a good idea to add some info logs (using `ALICEVISION_LOG_INFO`) to keep track of the process: I just tested it on a 2600 frames sequence and it took about 40 minutes, which without any logging can seem quite suspicious to users (as they might want at least a rough idea of how fast the process is going and how long it is going to last).

I have promoted some "DEBUG"-level logs to "INFO" for the smart selection, so we now have some info about the progress of the score computation with the default verbose setting.

simogasp commented on src/aliceVision/keyframe/README.md:
```diff
@@ -0,0 +1,135 @@
+
+Once both the sharpness and motion scores have been computed, subsequences are identified based on the motion accumulation across frames. The motion accumulation threshold is set with `pxDisplacement` which represents, in percent, the amount of pixels that need to have moved since the last keyframe for the motion to be significant. As the motion scores represent a displacement value for each fram ...
+
+Within each subsequence, a single frame is to be selected as a keyframe. Before processing to the selection itself, the number of identified subsequences is checked to ensure that the minimum and maximum number of requested output keyframes are respected.
```
```suggestion
Within each subsequence, a single frame is to be selected as a keyframe. Before processing the selection itself, the number of identified subsequences is checked to ensure that the minimum and maximum number of requested output keyframes are respected.
```

cbentejac replied:
Oops, I actually meant "proceeding to", thanks for catching that!

simogasp commented on src/aliceVision/keyframe/README.md:
```diff
@@ -0,0 +1,135 @@
+
+### Selection
+
+Once both the sharpness and motion scores have been computed, subsequences are identified based on the motion accumulation across frames. The motion accumulation threshold is set with `pxDisplacement` which represents, in percent, the amount of pixels that need to have moved since the last keyframe for the motion to be significant. As the motion scores represent a displacement value for each fram ...
```
```suggestion
Once both the sharpness and motion scores have been computed, subsequences are identified based on the motion accumulation across frames. The motion accumulation threshold is set with `pxDisplacement` which represents, in per cent, the number of pixels that need to have moved since the last keyframe for the motion to be significant. As the motion scores represent a displacement value for each frame, summing them over time until the accumulation reaches the threshold allows to divide the input video / sequence into subsequences that all contain significant motion.
```

cbentejac replied:
Done, thanks!
